core/vm/contracts.go

```diff
@@ -669,3 +675,50 @@ func (c *epochSize) Run(input []byte, caller common.Address, evm *EVM, gas uint6
 
 	return epochSizeBytes, gas, nil
 }
+
+type blockNumberFromHeader struct{}
+
+func (c *blockNumberFromHeader) RequiredGas(input []byte) uint64 {
+	return params.GetBlockNumberFromHeaderGas
+}
+
+func (c *blockNumberFromHeader) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
+	gas, err := debitRequiredGas(c, input, gas)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	var header types.Header
+	err = rlp.DecodeBytes(input, &header)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	blockNumber := header.Number.Bytes()
+	blockNumberBytes := common.LeftPadBytes(blockNumber[:], 32)
+
+	return blockNumberBytes, gas, nil
+}
+
+type hashHeader struct{}
+
+func (c *hashHeader) RequiredGas(input []byte) uint64 {
+	return params.HashHeaderGas
+}
+
+func (c *hashHeader) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
+	gas, err := debitRequiredGas(c, input, gas)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	var header types.Header
+	err = rlp.DecodeBytes(input, &header)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	hashBytes := header.Hash().Bytes()
```

Comment:
Assuming this is guaranteed to be 32 bytes?

Reply:
Yes, I believe so. I see no path (after rlp.DecodeBytes with nil error) that doesn't return 32 bytes. But I tested that the unit tests run regardless of the return value.  And I'll also test what Solidity would do with a non-32-byte return value.